source/bin/pr-log.ts

```diff
@@ -47,20 +49,24 @@ program
     .storeOptionsAsProperties(false)
     .description(config.description ?? '')
     .version(config.version ?? '')
-    .arguments('<version-number>')
-    .option('--sloppy', 'skip ensuring clean local git state')
-    .option('--trace', 'show stack traces for any error')
+    .argument('[version-number]', 'Desired version number. Optional when --unreleased was provided')
+    .option('--sloppy', 'skip ensuring clean local git state', false)
+    .option('--trace', 'show stack traces for any error', false)
     .option('--default-branch <name>', 'set custom default branch', 'main')
-    .option('--stdout', 'output the changelog to stdout instead of writing to CHANGELOG.md')
-    .action(async (versionNumber: string, options: Record<string, unknown>) => {
-        const runOptions: RunOptions = {
-            sloppy: options.sloppy === true,
-            changelogPath,
-            stdout: options.stdout === true
-        };
+    .option('--stdout', 'output the changelog to stdout instead of writing to CHANGELOG.md', false)
+    .option('--unreleased', 'include section for unreleased changes', false)
+    .hook('preAction', (thisCommand) => {
+        const { unreleased } = thisCommand.opts<Record<string, unknown>>();
+        const hasArguments = thisCommand.args.length > 0;
+
+        if (unreleased === true && hasArguments) {
+            throw new InvalidArgumentError('A version number is not allowed when --unreleased was provided');
```

Comment:
Isn’t this already handled by `createCliRunOptions()`?

Reply:
Correct. The only difference here is that this `preAction` hook is handled by [commander](https://github.com/tj/commander.js) whereas the other one is "self written".

Theoretically we could get rid of this commander hook here with the "disadvantage" that all other code (GitHub authentication, reading `package.json` and so on) is still running even when the arguments are invalid.

We could also remove our `createCliRunOptions()` but I wanted to have a compiler save discriminated union type in all other areas of the code base that ensures either one or the other state.

What do you think?

Reply:
Can’t we simply run `createCliRunOptions` as the first thing in `.action()` an everything else in the `Ok()` handler?

Reply:
Something like c7d2f5d911782d8d6ef8ff85817b085039b5a1bd ?